src/effects/VST/VSTEffect.cpp

```diff
@@ -1026,60 +1041,91 @@ auto VSTEffectMessage::Clone() const -> std::unique_ptr<Message>
    // Make sure of the chunk capacity
-   result->settings.mChunk.reserve(this->settings.mChunk.capacity());
+   result->mChunk.reserve(this->mChunk.capacity());
+
    return result;
 }
 
 void VSTEffectMessage::Assign(Message && src)
 {
-   auto& dstSettings = this->settings;
-   auto& srcSettings = static_cast<VSTEffectMessage&>(src).settings;
+   VSTEffectMessage& vstSrc = static_cast<VSTEffectMessage&>(src);
+
+   mChunk = vstSrc.mChunk;
+   vstSrc.mChunk.resize(0);     // capacity will be preserved though
+
+   assert(mParamsVec.size() == vstSrc.mParamsVec.size());
+
+   for (size_t i = 0; i < mParamsVec.size(); i++)
+   {
+      mParamsVec[i] = vstSrc.mParamsVec[i];
 
-   VSTEffectWrapper::MoveSettingsContents(
-      std::move(srcSettings), dstSettings, /*merge=*/false);
+      // consume the source value
+      vstSrc.mParamsVec[i] = std::nullopt;
+   }
 }
 
 void VSTEffectMessage::Merge(Message && src)
 {
-   auto& dstSettings = this->settings;
+   VSTEffectMessage& vstSrc = static_cast<VSTEffectMessage&>(src);
 
-   auto& srcSettings = static_cast<VSTEffectMessage&>(src).settings;
+   bool chunkWasAssigned = false;
 
+   if ( ! vstSrc.mChunk.empty() )
+   {
+      mChunk = vstSrc.mChunk;
+      chunkWasAssigned = true;
+   }
+
+   vstSrc.mChunk.resize(0);  // capacity will be preserved though
+
+   assert(mParamsVec.size() == vstSrc.mParamsVec.size());
+
+   for (size_t i = 0; i < mParamsVec.size(); i++)
+   {
+      if (chunkWasAssigned)
+      {
+         mParamsVec[i] = vstSrc.mParamsVec[i];
+      }
+      else
+      {
+         // if src val is nullopt, do not copy it to dest
+         if (vstSrc.mParamsVec[i] != std::nullopt)
+         {
+            mParamsVec[i] = vstSrc.mParamsVec[i];
+         }
+      }
+
+      // consume the source value
+      vstSrc.mParamsVec[i] = std::nullopt;
+   }
 
-   VSTEffectWrapper::MoveSettingsContents(
-      std::move(srcSettings), dstSettings, /*merge=*/true);
 }
 
 
 std::unique_ptr<EffectInstance::Message> VSTEffectInstance::MakeMessage() const
 {
    VSTEffectSettings settings;
    FetchSettings(settings, false);
-   return std::make_unique<VSTEffectMessage>(std::move(settings));
-}
 
-
-std::unique_ptr<EffectInstance::Message> VSTEffectInstance::MakeMessage(int id, double value) const
-{
-   VSTEffectSettings settings;
+   VSTEffectMessage::ParamVector paramVector;
+   paramVector.resize(settings.mParamsMap.size(), std::nullopt);
 
    ForEachParameter
    (
       [&](const ParameterInfo& pi)
       {
-         if (pi.mID == id)
-         {
-            settings.mParamsMap[pi.mName] = { id, value };
-         }
-         else
-         {
-            settings.mParamsMap[pi.mName] = std::nullopt;
-         }
+         paramVector[pi.mID] = settings.mParamsMap[pi.mName];
          return true;
       }
    );
 
-   return std::make_unique<VSTEffectMessage>(std::move(settings));
+   return std::make_unique<VSTEffectMessage>(settings.mChunk, paramVector);
```

Comment:
`move()` both arguments to avoid vectory copies

Reply:
ok